src/Components/Components/src/DynamicComponent.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// A component that renders another component dynamically according to its
+    /// <see cref="Type" /> parameter.
+    /// </summary>
+    public class DynamicComponent : IComponent
+    {
+        private RenderHandle _renderHandle;
+        private RenderFragment _cachedRenderFragment;
+
+        /// <summary>
+        /// Constructs an instance of <see cref="DynamicComponent"/>.
+        /// </summary>
+        public DynamicComponent()
+        {
+            _cachedRenderFragment = Render;
+        }
+
+        /// <summary>
+        /// Gets or sets the type of the component to be rendered. The supplied type must
+        /// implement <see cref="IComponent"/>.
```

Comment:
Should we explicitly check and throw a clear exception if Type doesn't implement `IComponent`?

Reply:
The renderer does that anyway, so I didn't want to duplicate the behavior and pay the cost of checking it twice. The error message from the renderer is very clear:

```
$"The type {componentType.FullName} does not implement {nameof(IComponent)}."
```

Reply:
Cool didn't know that, thanks 👍